src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java

```diff
@@ -363,54 +363,144 @@ private String breadcrumbs() {
+
+        // Assert invariants
+        public SequenceElementByIndex {
+            if (index < 0) {
+                throw new IllegalArgumentException("Index < 0");
             }
+        }
+
+        @Override
+        public LayoutPath apply(LayoutPath layoutPath) {
+            return layoutPath.sequenceElement(index);
+        }
+
+        @Override
+        public String toString() {
+            return "sequenceElement(" + index + ")";
+        }
+
+    }
+
+    public record SequenceElementByRange(long start, long step)
+            implements MemoryLayout.PathElement, UnaryOperator<LayoutPath> {
 
-            public String description() {
-                return description;
+        // Assert invariants
+        public SequenceElementByRange {
+            if (start < 0) {
+                throw new IllegalArgumentException("Start index must be positive: " + start);
+            }
+            if (step == 0) {
+                throw new IllegalArgumentException("Step must be != 0: " + step);
             }
         }
 
-        final PathKind kind;
-        final UnaryOperator<LayoutPath> pathOp;
-        final String stringRepresentation;
+        @Override
+        public LayoutPath apply(LayoutPath layoutPath) {
+            return layoutPath.sequenceElement(start, step);
+        }
+
+        @Override
+        public String toString() {
+            return "sequenceElement(" + start + ", " + step + ")";
+        }
+
+    }
+
+    public record SequenceElement()
+            implements MemoryLayout.PathElement, UnaryOperator<LayoutPath> {
+
+        private static final SequenceElement INSTANCE = new SequenceElement();
+
+        @Override
+        public LayoutPath apply(LayoutPath layoutPath) {
+            return layoutPath.sequenceElement();
+        }
+
+        @Override
+        public int hashCode() {
+            return 31;
+        }
 
-        public PathElementImpl(PathKind kind,
-                               UnaryOperator<LayoutPath> pathOp,
-                               String stringRepresentation) {
-            this.kind = kind;
-            this.pathOp = pathOp;
-            this.stringRepresentation = stringRepresentation;
+        @Override
+        public String toString() {
+            return "sequenceElement()";
+        }
+
+        public static MemoryLayout.PathElement instance() {
+            return INSTANCE;
         }
 
+    }
+
+    public record DereferenceElement()
+            implements MemoryLayout.PathElement, UnaryOperator<LayoutPath> {
+
+        private static final DereferenceElement INSTANCE = new DereferenceElement();
+
         @Override
         public LayoutPath apply(LayoutPath layoutPath) {
-            return pathOp.apply(layoutPath);
+            return layoutPath.derefElement();
         }
 
-        public PathKind kind() {
-            return kind;
+        @Override
+        public int hashCode() {
+            return 63;
         }
 
```

Comment:
```suggestion
```

Reply:
So, a record with no component will have a `hashCode()` of zero.  I didn't want the two singleton variants to have the same hash code. I've added a test for this and will remove one of the overrides.